plugins/package-managers/node/src/main/kotlin/pnpm/ModuleInfo.kt

```diff
@@ -19,12 +19,20 @@
 
 package org.ossreviewtoolkit.plugins.packagemanagers.node.pnpm
 
+import java.io.ByteArrayInputStream
+
 import kotlinx.serialization.Serializable
+import kotlinx.serialization.json.DecodeSequenceMode
 import kotlinx.serialization.json.Json
+import kotlinx.serialization.json.decodeToSequence
 
 private val JSON = Json { ignoreUnknownKeys = true }
 
-internal fun parsePnpmList(json: String): List<ModuleInfo> = JSON.decodeFromString(json)
+internal fun parsePnpmList(json: String): List<ModuleInfo> =
+    JSON.decodeToSequence<List<ModuleInfo>>(
+        ByteArrayInputStream(json.toByteArray()),
+        DecodeSequenceMode.WHITESPACE_SEPARATED
+    ).toList().flatten()
```

Comment:
Hm, looking at [1] it seems that "nested project" is just a project which is in a descendant dir of another project.
If ORT analyzes the parent dir or any ancestor, IIUC it would find both definition files / projects and analyze them separately. 

Calling flatten, as above, I'd expect the dependencies of the nested project to become dependencies of the outer one. Why is this intended?

(I wonder if instead of `flatten()`,  `first()` has to be called. (So that the inner project gets ignored, as it will be analyzed separately anyway) 


[1] https://github.com/oss-review-toolkit/ort/tree/fbb3a47bb65f25c0b67c4d312cd0adb3e29960c9/plugins/package-managers/node/src/funTest/assets/projects/synthetic/pnpm/nested-project

Reply:
> Hm, looking at [1] it seems that "nested project" is just a project which is in a descendant dir of another project.
> If ORT analyzes the parent dir or any ancestor, IIUC it would find both definition files / projects and analyze them separately.

BTW, some analyzers already have code to filter out definition files from nested vendor directories, which is a similar case, also see https://github.com/oss-review-toolkit/ort/pull/9864.

So maybe also for PNPM, nested definition file should be filtered out, as the `pnpm` CLI takes care by itself to regard them?

Reply:
> BTW, some analyzers already have code to filter out definition files from nested vendor directories, which is a similar case, also see #9864.

Can you give a hint why you think this is a similar case? (I wonder whether these projects are actually independent besides being in subdirs)

> So maybe also for PNPM, nested definition file should be filtered out, as the pnpm CLI takes care by itself to regard them?

There is also the option of disregarding the nested subproject wrt to a definition file `resolveDependencies()`.

Note: For workspaces, it is necessary to skip the analysis of the individual definition files and analyze them from the workspace root in some cases, because the workspace root is where the _lockfile_ resides. How does the lockfile situation look here?

Reply:
I have run tests with both approaches, using `flatten()` and `first()`, against the real project that was failing before. The results are almost identical; the only difference is that in case of `flatten()` the nested project shows up in the list of projects in the ORT result, while this is not the case with `first()`.

So, what do you prefer? For me both options would be fine as long as the Analyzer run is successful.

On a side note, is it safe to use `first()`? Can we assume that there is a defined order in the output of PNPM?

Reply:
> So, what do you prefer?

I prefer to analyze the nested project separately, if possible (consider lockfile location), and thus use first / single approach.

> On a side note, is it safe to use `first()`? Can we assume that there is a defined order in the output of PNPM?

Note sure, but it certainly should be possible to find the `single` item which corresponds to the project in the current working dir.

Reply:
> Can you give a hint why you think this is a similar case?

It's explained in the next sentence that I wrote: "as the pnpm CLI takes care by itself to regard them?"

If you have `root/package.json` and `root/nested/package.json`, and `pnpm list` itself would already regard `root/nested/package.json` when run on `root/package.json`, then probably `root/nested/package.json` should get filtered out to avoid running `pnpm list` on it separately. And that's a similar case to vendored dependencies if you just swap out "nested" for "vendor".

But that was just me thinking out aloud.

In any case the change looks sensible to me, as it teaches to handle a special case of `pnpm list` output correctly (independently of what we do with that information), which by itself is good.

Reply:
> On a side note, is it safe to use `first()`? Can we assume that there is a defined order in the output of PNPM?

Good point. I tend to prefer `flatten()` to make that function more generic and not lose any information at that point already. Eventually, callers of that function should do any filtering.

Reply:
The preferences of the reviewers are somewhat contradictory with regard to the `flatten()` vs `first()` approach.
Would the following be an approach everybody can agree with:
* `parsePnpmList()` just returns a `Sequence<List<ModuleInfo>>`. Then no information is lost at this point.
* `PNPM` filters the sequence for the list of paths of the top-level project and processes only those.

@fviernau, it is not clear to me based on which criteria this filtering can be done. How can the list that corresponds to the definition file be identified? If there is only a single path involved, I would go for the shortest path, but if there are multiple, I am not sure how to handle this.

Reply:
> Would the following be an approach everybody can agree with:

That would work for me.

Reply:
> Regarding the sequence return type. Does it safe any JSON parsing?

The docs of `decodeToSequence` indicate so: "Transforms the given stream into *lazily* deserialized sequence of elements of type T".